src/KGBotka/Calc.hs

```diff
@@ -0,0 +1,210 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE LambdaCase #-}
+
+module KGBotka.Calc where
+
+import Control.Applicative (Alternative(..))
+import Control.Monad (void)
+import Control.Monad.Trans.Class (lift)
+import Control.Monad.Trans.Except (ExceptT(..), throwE)
+import Data.Char (isAlpha, isDigit)
+import qualified Data.Map.Strict as M
+import qualified Data.Text as T
+import KGBotka.Parser
+import System.Random (randomIO)
+
+data Operator
+  = Add
+  | Sub
+  | Mul
+  | Div
+  | Mod
+  | Pow
+  deriving (Show, Eq)
+
+charToOperator :: Char -> Operator
+charToOperator '+' = Add
+charToOperator '-' = Sub
+charToOperator '*' = Mul
+charToOperator '/' = Div
+charToOperator '%' = Mod
+charToOperator '^' = Pow
+charToOperator _ =
+    -- TODO: Make charToOperator not throw an exception
+    -- For now it is an unsafe function because it indicates
+    -- that operators were added but the parsers were not
+    -- corrected.
+  error
+    "charToOperator: No pattern for operator conversion. \
+    \Maybe you added a new operator and forgot to add a case to the conversion function."
+
+data CalcExpression
+  = BinaryExpression Operator CalcExpression CalcExpression
+  | NegativeExpression CalcExpression
+  | FunctionApplication T.Text [CalcExpression]
+  | ValueExpression Double
+  deriving (Show)
+
+newtype CalcEvalError =
+  CalcEvalError T.Text
+  deriving (Show)
+
+type CalcEval = ExceptT CalcEvalError IO
```

Comment:
I don't like the fact that the only reason we need ExceptT with IO here is because we use a single function that returns IO (randomIO). The ratio of complexity increase/benefit is quite bad. 

On the other hand this is more of a fault of Haskell than the implementation, so I'll pass.